### PR TITLE
19665 - update alternate name model

### PR DESCRIPTION
**Pull request by tshyun24**
*Issue #:* /bcgov/entity#19665

*Description of changes:*
- Add `mailing_address_id`, `delivery_address_id` and `email` to alternate name model
- Add office address properties to alternate names model.
- Add `alternate_name_id` to offices model
-  Update alembic script to add alternate_name_id to addresses table and add FKs and indexes as appropriate
-  Add alternate_id to addresses model and add relationships as appropriate

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

**Review by vysakh-menon-aot — APPROVED** [review]

File: legal-api/src/legal_api/models/alternate_name.py [hunk 1]
```diff
@@ -111,6 +119,56 @@ def find_by_name(cls, name: str = None):
         alternate_name = cls.query.filter_by(name=name).filter_by(end_date=None).one_or_none()
         return alternate_name
 
+    @property
+    def office_mailing_address(self):
+        """Return the mailing address."""
+        registered_office = (
+            db.session.query(Office)
+            .filter(Office.legal_entity_id == self.id)
```
Comment (argush3) [inline legal-api/src/legal_api/models/alternate_name.py:127]: should be on `Office.alternate_entity_id`

File: legal-api/src/legal_api/models/alternate_name.py [hunk 2]
```diff
@@ -111,6 +119,56 @@ def find_by_name(cls, name: str = None):
...
+            .filter(Office.office_type == "businessOffice")
+            .one_or_none()
+        ):
+            return business_office.addresses.filter(Address.address_type == "mailing")
+
+        return (
+            db.session.query(Address)
+            .filter(Address.legal_entity_id == self.id)
+            .filter(Address.address_type == Address.MAILING)
+        )
```
Comment (argush3) [inline legal-api/src/legal_api/models/alternate_name.py:145]: i'm not sure about this.  i don't know if we actually save addresses directly to the LE right now. 

Going to check prod data.

File: legal-api/src/legal_api/models/alternate_name.py [hunk 3]
```diff
@@ -111,6 +119,56 @@ def find_by_name(cls, name: str = None):
...
+    @property
+    def office_mailing_address(self):
+        """Return the mailing address."""
+        registered_office = (
+            db.session.query(Office)
+            .filter(Office.legal_entity_id == self.id)
+            .filter(Office.office_type == "registeredOffice")
+            .one_or_none()
+        )
+        if registered_office:
```
Comment (vysakh-menon-aot) [inline legal-api/src/legal_api/models/alternate_name.py:131]: If AlternateName contains only SP and GP then it won't have any other office_type than businessOffice

File: legal-api/src/legal_api/models/alternate_name.py (hunk 2 above)
Comment (argush3) [reply legal-api/src/legal_api/models/alternate_name.py:145]: so it seems like when we dissolve a business right now, it creates the custodian address which differs from how we have saved other office addresses.  As in it populates both the `legal_entity_id` and the `office_id` in the `addresses` table.

For now, I think we should just add the `alternate_name_id` to `addresses` table too.  This way we can just follow the existing pattern for SPs

File: legal-api/src/legal_api/models/alternate_name.py (hunk 1 above)
Comment (chenhongjing) [reply legal-api/src/legal_api/models/alternate_name.py:127]: More correctly, should be `Office.alternate_name_id` 😄

File: legal-api/src/legal_api/models/alternate_name.py [hunk 4]
```diff
@@ -111,6 +119,56 @@ def find_by_name(cls, name: str = None):
...
+            .filter(Address.legal_entity_id == self.id)
+            .filter(Address.address_type == Address.MAILING)
+        )
+
+    @property
+    def office_delivery_address(self):
+        """Return the delivery address."""
+        registered_office = (
+            db.session.query(Office)
+            .filter(Office.legal_entity_id == self.id)
```
Comment (chenhongjing) [inline legal-api/src/legal_api/models/alternate_name.py:152]: Remember to change this to `Office.alternate_name_id` too 😺

File: legal-api/src/legal_api/models/alternate_name.py (hunk 1 above)
Comment (chenhongjing) [reply legal-api/src/legal_api/models/alternate_name.py:127]: More correctly, should be `Office.alternate_name_id` 😄

File: legal-api/src/legal_api/models/alternate_name.py [hunk 5]
```diff
@@ -111,6 +119,40 @@ def find_by_name(cls, name: str = None):
...
+            business_office := db.session.query(Office)  # SP/GP
+            .filter(Office.alternate_name_id == self.id)
+            .filter(Office.office_type == "businessOffice")
+            .one_or_none()
+        ):
+            return business_office.addresses.filter(Address.address_type == "mailing")
+
+        return (
+            db.session.query(Address)
+            .filter(Address.legal_entity_id == self.id)
```
Comment (argush3) [inline legal-api/src/legal_api/models/alternate_name.py:135]: `Address.legal_entity_id` should be `Address.alternate_name_id`

File: legal-api/src/legal_api/models/alternate_name.py [hunk 6]
```diff
@@ -111,6 +119,40 @@ def find_by_name(cls, name: str = None):
...
+            business_office := db.session.query(Office)  # SP/GP
+            .filter(Office.alternate_name_id == self.id)
+            .filter(Office.office_type == "businessOffice")
+            .one_or_none()
+        ):
+            return business_office.addresses.filter(Address.address_type == "delivery")
+
+        return (
+            db.session.query(Address)
+            .filter(Address.legal_entity_id == self.id)
```
Comment (argush3) [inline legal-api/src/legal_api/models/alternate_name.py:152]: `Address.legal_entity_id` should be `Address.alternate_name_id`

File: legal-api/migrations/versions/004f7558063f_second.py [hunk 7]
```diff
@@ -32,9 +32,11 @@ def upgrade():
     sa.Column('change_filing_id', sa.Integer(), nullable=True),
     sa.Column('office_id', sa.Integer(), nullable=True),
     sa.Column('version', sa.Integer(), nullable=False),
+    sa.Column('alternate_name_id', sa.Integer(), nullable=True),
```
Comment (argush3) [inline legal-api/migrations/versions/004f7558063f_second.py:35]: you'll need to probably do all the same upgrade/downgrade updates you did in the alembic script for  `addresses` table to the `addresses_history` table

File: legal-api/src/legal_api/models/alternate_name.py (hunk 5 above)
Comment (tshyun24) [reply legal-api/src/legal_api/models/alternate_name.py]: Thank you for catching that!

File: legal-api/migrations/versions/004f7558063f_second.py [hunk 8]
```diff
@@ -1281,6 +1287,7 @@ def downgrade():
         batch_op.drop_index(batch_op.f('ix_addresses_legal_entity_id'))
         batch_op.drop_index(batch_op.f('ix_addresses_change_filing_id'))
         batch_op.drop_index(batch_op.f('ix_addresses_address_type'))
+        batch_op.drop_index(batch_op.f('ix_addresses_alternate_name_id'))
```
Comment (argush3) [inline legal-api/migrations/versions/004f7558063f_second.py:1290]: need to drop `ix_addresses_alternate_name_id ` index for `addresses_history` table too

File: legal-api/migrations/versions/004f7558063f_second.py (hunk 7 above)
Comment (argush3) [reply legal-api/migrations/versions/004f7558063f_second.py:35]: that would be included in #19429 i think

**Pull request merged**